test/001_highsnr_stdio_audio/test_callback_rx_outside.py

```diff
@@ -0,0 +1,181 @@
+        if self.AUDIO_INPUT_DEVICE != -1: 
+            self.p = pyaudio.PyAudio()
+            # auto search for loopback devices
+            if self.AUDIO_INPUT_DEVICE == -2:
+                loopback_list = []
+                for dev in range(0,self.p.get_device_count()):
+                    if 'Loopback: PCM' in self.p.get_device_info_by_index(dev)["name"]:
+                        loopback_list.append(dev)
+                if len(loopback_list) >= 2:
+                    self.AUDIO_INPUT_DEVICE = loopback_list[0] #0  = RX   1 = TX
+                    print(f"loopback_list rx: {loopback_list}", file=sys.stderr)
+                else:
+                    quit()
+                    
+            print(f"AUDIO INPUT DEVICE: {self.AUDIO_INPUT_DEVICE} DEVICE: {self.p.get_device_info_by_index(self.AUDIO_INPUT_DEVICE)['name']}  \
+                    AUDIO SAMPLE RATE: {self.AUDIO_SAMPLE_RATE_RX}", file=sys.stderr)
+            
+            self.stream_rx = self.p.open(format=pyaudio.paInt16, 
+                                    channels=1,
+                                    rate=self.AUDIO_SAMPLE_RATE_RX,
+                                    frames_per_buffer=self.AUDIO_FRAMES_PER_BUFFER,
+                                    input=True,
+                                    output=False,
+                                    input_device_index=self.AUDIO_INPUT_DEVICE,
+                                    stream_callback=self.callback
+                                    ) 
+
+        # open codec2 instance        
+        self.freedv = cast(codec2.api.freedv_open(self.MODE), c_void_p)
+        
+        # get number of bytes per frame for mode
+        self.bytes_per_frame = int(codec2.api.freedv_get_bits_per_modem_frame(self.freedv)/8)
+        
+        self.bytes_out = create_string_buffer(self.bytes_per_frame * 2)
+        
+        codec2.api.freedv_set_frames_per_burst(self.freedv,self.N_FRAMES_PER_BURST)
+        
+        self.total_n_bytes = 0
+        self.rx_total_frames = 0
+        self.rx_frames = 0
+        self.rx_bursts = 0
+        self.rx_errors = 0
+        self.nread_exceptions = 0
+        self.timeout = time.time() + self.TIMEOUT
+        self.receive = True
+        self.audio_buffer = codec2.audio_buffer(self.AUDIO_FRAMES_PER_BUFFER*2)
+        self.resampler = codec2.resampler()
+        
+        # Copy received 48 kHz to a file.  Listen to this file with:
+        #   aplay -r 48000 -f S16_LE rx48_callback.raw
+        # Corruption of this file is a good way to detect audio card issues
+        self.frx = open("rx48_callback.raw", mode='wb')    
+           
+    def callback(self, data_in48k, frame_count, time_info, status):
+        
+        x = np.frombuffer(data_in48k, dtype=np.int16)
+        x.tofile(self.frx)
+        x = self.resampler.resample48_to_8(x)    
+        self.audio_buffer.push(x)
+
+        return (None, pyaudio.paContinue)
```

Comment:
@drowe67 I created this test to determine if we have any differences with decoding inside and outside of the callback. This could become interesting for multimode decoding. 
It could also be possible, we have less memory address problems because of the class :thinking:

Reply:
I'm sure it's possible either way, however with the decoding outside the callback we will need way to prevent two threads accessing the audio buffer at the same time (this will eventually happen with a long enough run).